src/main/java/org/elasticsearch/script/Script.java

```diff
@@ -19,53 +19,83 @@
 
 package org.elasticsearch.script;
 
-import java.util.Map;
-
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Streamable;
+import org.elasticsearch.common.logging.support.LoggerMessageFormat;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.script.ScriptService.ScriptType;
 
-import static org.elasticsearch.script.ScriptService.ScriptType;
+import java.io.IOException;
+import java.util.Map;
 
 /**
  * Script holds all the parameters necessary to compile or find in cache and then execute a script.
  */
-public class Script {
+public class Script implements ToXContent, Streamable {
 
-    private final String lang;
-    private final String script;
-    private final ScriptType type;
-    private final Map<String, Object> params;
+    public static final ScriptType DEFAULT_TYPE = ScriptType.INLINE;
+    private static final ScriptParser PARSER = new ScriptParser();
+
+    private String script;
+    private @Nullable ScriptType type;
+    private @Nullable String lang;
+    private @Nullable Map<String, Object> params;
+
+    /**
+     * For Serialization
+     */
+    Script() {
```

Comment:
we can use Writeable here instead so fields can become final and default constructor can go away